File: src/csharp/Foundgine.Runtime/ControlPlane/Approvals/IApprovalStore.cs
namespace Foundgine.Runtime.ControlPlane.Approvals;

public interface IApprovalStore
{
    ApprovalRequest Create(string requestFingerprint, int requiredApprovals = 1);

    bool TryGet(string approvalId, out ApprovalRequest? request);

    ApprovalRequest RecordDecision(string approvalId, ApprovalGrant grant);
}

/// <summary>
/// Process-local approval store. A production deployment with real
/// human-in-the-loop approvers should back <see cref="IApprovalStore"/>
/// with durable storage — pending approvals must survive a process
/// restart — but the interface and workflow shape stay the same.
/// </summary>
public sealed class InMemoryApprovalStore : IApprovalStore
{
    private readonly Dictionary<string, ApprovalRequest> _requests = new(StringComparer.Ordinal);
    private readonly Lock _gate = new();

    public ApprovalRequest Create(string requestFingerprint, int requiredApprovals = 1)
    {
        var request = ApprovalRequest.Create(Guid.NewGuid().ToString("n"), requestFingerprint, requiredApprovals);
        lock (_gate)
        {
            _requests[request.ApprovalId] = request;
        }

        return request;
    }

    public bool TryGet(string approvalId, out ApprovalRequest? request)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(approvalId);
        lock (_gate)
        {
            return _requests.TryGetValue(approvalId, out request);
        }
    }

    public ApprovalRequest RecordDecision(string approvalId, ApprovalGrant grant)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(approvalId);
        ArgumentNullException.ThrowIfNull(grant);

        lock (_gate)
        {
            if (!_requests.TryGetValue(approvalId, out var existing))
                throw new KeyNotFoundException($"No approval request '{approvalId}' exists.");

            var updated = existing.WithGrant(grant);
            _requests[approvalId] = updated;
            return updated;
        }
    }
}

File: src/java/foundgine-runtime/src/main/java/com/foundgine/runtime/controlplane/approvals/IApprovalStore.java
package com.foundgine.runtime.controlplane.approvals;

import java.util.Optional;

/**
 * Port of {@code Foundgine.Runtime.ControlPlane.Approvals.IApprovalStore}.
 *
 * <p>
 * <b>Porting decision:</b> the C# member
 * {@code bool TryGet(string, out ApprovalRequest?)} has no direct Java
 * equivalent (Java has no {@code out} parameters). It is ported as
 * {@link #tryGet(String)} returning {@code Optional<ApprovalRequest>}; an empty
 * {@code Optional} corresponds to the C# method returning {@code false}.
 */
public interface IApprovalStore {
	ApprovalRequest create(String requestFingerprint, int requiredApprovals);

	default ApprovalRequest create(String requestFingerprint) {
		return create(requestFingerprint, 1);
	}

	Optional<ApprovalRequest> tryGet(String approvalId);

	ApprovalRequest recordDecision(String approvalId, ApprovalRequest.ApprovalGrant grant);
}
